algebra/src/curves/tweedle/tests.rs

```diff
@@ -0,0 +1,177 @@
+
+#[test]
+fn test_dee_projective_curve() {
+    curve_tests::<dee::Projective>();
+}
+
+#[test]
+fn test_dee_projective_group() {
+    let mut rng = XorShiftRng::seed_from_u64(1234567890u64);
+    let a: dee::Projective = rng.gen();
+    let b: dee::Projective = rng.gen();
+    group_test(a, b);
+}
+
+#[test]
+fn test_dee_generator() {
+    let generator = dee::Affine::prime_subgroup_generator();
+    assert!(generator.is_on_curve());
+    assert!(generator.is_in_correct_subgroup_assuming_on_curve());
+}
+
+#[test]
+fn test_dum_projective_curve() {
+    curve_tests::<dum::Projective>();
+}
+
+#[test]
+fn test_dum_projective_group() {
+    let mut rng = XorShiftRng::seed_from_u64(1234567890u64);
+    let a: dum::Projective = rng.gen();
+    let b: dum::Projective = rng.gen();
+    group_test(a, b);
+}
+
+#[test]
+fn test_dum_generator() {
+    let generator = dum::Affine::prime_subgroup_generator();
+    assert!(generator.is_on_curve());
+    assert!(generator.is_in_correct_subgroup_assuming_on_curve());
+}
+
+#[test]
+fn test_dee_generator_raw() {
+    let mut x = Fq::zero();
+    let mut i = 0;
+    loop {
+        // y^2 = x^3 + b
+        let mut rhs = x;
+        rhs.square_in_place();
+        rhs.mul_assign(&x);
+        rhs.add_assign(&dee::TweedledeeParameters::COEFF_B);
+
+        if let Some(y) = rhs.sqrt() {
+            let p = dee::Affine::new(x, if y < -y { y } else { -y }, false);
+            assert!(p.is_in_correct_subgroup_assuming_on_curve());
+
+            let dee = p.scale_by_cofactor();
+            assert_eq!(dee.into_affine(), p);
+
+            if !dee.is_zero() {
+                assert_eq!(i, 1);
+                let dee = dee::Affine::from(dee);
+
+                assert!(dee.is_in_correct_subgroup_assuming_on_curve());
+
+                assert_eq!(dee, dee::Affine::prime_subgroup_generator());
+                break;
+            }
+        }
+
+        i += 1;
+        x.add_assign(&Fq::one());
+    }
+}
+
+#[test]
+fn test_dum_generator_raw() {
+    let mut x = Fr::zero();
+    let mut i = 0;
+    loop {
+        // y^2 = x^3 + b
+        let mut rhs = x;
+        rhs.square_in_place();
+        rhs.mul_assign(&x);
+        rhs.add_assign(&dum::TweedledumParameters::COEFF_B);
+
+        if let Some(y) = rhs.sqrt() {
+            let p = dum::Affine::new(x, if y < -y { y } else { -y }, false);
+            assert!(p.is_in_correct_subgroup_assuming_on_curve());
+
+            let dum = p.scale_by_cofactor();
+            assert_eq!(dum.into_affine(), p);
+
+            if !dum.is_zero() {
+                assert_eq!(i, 1);
+                let dum = dum::Affine::from(dum);
+
+                assert!(dum.is_in_correct_subgroup_assuming_on_curve());
+
+                assert_eq!(dum, dum::Affine::prime_subgroup_generator());
+                break;
+            }
+        }
+
+        i += 1;
+        x.add_assign(&Fr::one());
+    }
+}
+
+#[test]
+fn test_dee_addition_correctness() {
```

Comment:
I assume that the test vectors are from Halo?

Reply:
No, tests are self-generated using Maple as software to check the expected result